jwst/stpipe/utilities.py

```diff
@@ -144,3 +146,53 @@ def folder_traverse(folder_path, basename_regex='.+', path_exclude_regex='^$'):
         for file in files:
             if basename_regex.match(file):
                 yield os.path.join(root, file)
+
+
+def record_step_status(datamodel, cal_step, success=True):
+    """Record whether or not a step completed in meta.cal_step
+
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to modify in place
+
+    cal_step : str
+        The attribute in meta.cal_step for recording the status of the step
+
+    success : bool
+        If True, then 'COMPLETE' is recorded.  If False, then 'SKIPPED'
+    """
+    if success:
+        status = 'COMPLETE'
+    else:
+        status = 'SKIPPED'
+
+    if isinstance(datamodel, Sequence):
+        for model in datamodel:
+            model.meta.cal_step._instance[cal_step] = status
+    else:
+        datamodel.meta.cal_step._instance[cal_step] = status
+
+    # TODO: standardize cal_step naming to point to the official step name
+
+
+def query_step_status(datamodel, cal_step):
+    """Query the status of a step in meta.cal_step
+    
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to check
+        
+    cal_step : str
+        The attribute in meta.cal_step to check
+    
+    Returns
+    -------
+    status : str
+        The status of the step in meta.cal_step, typically 'COMPLETE' or 'SKIPPED'
+    """
+    if isinstance(datamodel, Sequence):
+        return getattr(datamodel[0].meta.cal_step, cal_step, NOT_SET)
```

Comment:
This may not be the right behavior for all composite models -- it might sometimes be necessary to skip a step for one model in the step, but not all of them.

I think it might be better to return COMPLETE if any model in the container has that as its status.  If there are no COMPLETEs, but there are SKIPPEDs, return SKIPPED.  Return NOT_SET only if there are no statuses set at all.

Reply:
that's a good point- will change

Reply:
What about dropping the `Sequence` support and just support a `DataModel`? That way the step (or whatever calling code) can index into the container in whatever way makes the most sense.

Reply:
The problem is that it's sometimes the top-level pipeline that needs to know the output status and it may not know enough about the return structure from a particular step to pick out the right datamodel from a container.